connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -16,56 +16,325 @@
+    ) {
+        Objects.requireNonNull(loggingContext);
+        Objects.requireNonNull(connectorOffsetsTopic);
+        Objects.requireNonNull(connectorStoreAdmin);
+        return new ConnectorOffsetBackingStore(
+                Time.SYSTEM,
+                loggingContext,
+                connectorOffsetsTopic,
+                null,
+                connectorStore,
+                connectorStoreAdmin
+        );
+    }
+
+    private final Time time;
+    private final Supplier<LoggingContext> loggingContext;
     private final String primaryOffsetsTopic;
+    private final Optional<OffsetBackingStore> workerStore;
+    private final Optional<KafkaOffsetBackingStore> connectorStore;
+    private final Optional<TopicAdmin> connectorStoreAdmin;
 
-    public ConnectorOffsetBackingStore(
+    ConnectorOffsetBackingStore(
+            Time time,
+            Supplier<LoggingContext> loggingContext,
+            String primaryOffsetsTopic,
             OffsetBackingStore workerStore,
-            String primaryOffsetsTopic
+            KafkaOffsetBackingStore connectorStore,
+            TopicAdmin connectorStoreAdmin
     ) {
-        this.workerStore = workerStore;
+        if (workerStore == null && connectorStore == null) {
+            throw new IllegalArgumentException("At least one non-null offset store must be provided");
+        }
+        this.time = time;
+        this.loggingContext = loggingContext;
         this.primaryOffsetsTopic = primaryOffsetsTopic;
+        this.workerStore = Optional.ofNullable(workerStore);
+        this.connectorStore = Optional.ofNullable(connectorStore);
+        this.connectorStoreAdmin = Optional.ofNullable(connectorStoreAdmin);
     }
 
     public String primaryOffsetsTopic() {
         return primaryOffsetsTopic;
     }
 
+    /**
+     * If configured to use a connector-specific offset store, {@link OffsetBackingStore#start() start} that store.
+     *
+     * <p>The worker-global offset store is not modified; it is the caller's responsibility to ensure that it is started
+     * before calls to {@link #get(Collection)} and {@link #set(Map, Callback)} take place.
+     */
     @Override
     public void start() {
-        // TODO
+        // Worker offset store should already be started
+        connectorStore.ifPresent(OffsetBackingStore::start);
     }
 
+    /**
+     * If configured to use a connector-specific offset store, {@link OffsetBackingStore#start() stop} that store,
+     * and the {@link TopicAdmin} used by that store.
+     *
+     * <p>The worker-global offset store is not modified as it may be used for other connectors that either already exist,
+     * or will be created, on this worker.
+     */
     @Override
     public void stop() {
-        // TODO
+        // Worker offset store should not be stopped as it may be used for multiple connectors
+        connectorStore.ifPresent(OffsetBackingStore::stop);
+        connectorStoreAdmin.ifPresent(TopicAdmin::close);
```

Comment:
Given that we close the `TopicAdmin` for source tasks in `AbstractWorkerSourceTask::close` [here](https://github.com/apache/kafka/blob/1e21201ea24389bdaccb8a462f3a53e356b58a58/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java#L306-L308), this may seem redundant. However, I've chosen to keep it in for two reasons:
1. The `WorkerConnector` class has no access to the `TopicAdmin` used for its offset store; we'd have to add another constructor parameter to that class and manually close the `TopicAdmin` during shutdown if we removed this line
2. It seems safer to leave this in. Although `Autocloseable::close` is not required to be idempotent, it is strongly recommended, and since all current calls to `ConnectorOffsetBackingStore::stop` are wrapped in `Utils::closeQuietly` anyways, the risks are low. On the other hand, by leaving this out, we would make it easier to introduce a resource leak in the future, which would have more serious consequences.